File: matlab_files/diff_calculator.m
function [figure_counter] = diff_calculator(diff_matrix_path_incomplete,min_lat,max_lat,...
    min_lon,max_lon,run,depths_to_plot,iteration,step,figure_counter)
% Calculate the differences between two similar plots, where possible.
% [1 0 1 145 150]
% [1 0 2 145 150]

resolution = 0.25;
latlim = [min_lat max_lat];
lonlim = [min_lon max_lon];
min_depth = depths_to_plot(1);
max_depth = depths_to_plot(2);
s = referenceSphere('Earth');
lat_lin = max_lat:-resolution:min_lat;
lon_lin = min_lon:resolution:max_lon;
[gridded_lon,gridded_lat] = meshgrid(lon_lin,lat_lin);
% plot_lon = reshape(gridded_lon, [numel(gridded_lon),1]);
% plot_lat = reshape(gridded_lat, [numel(gridded_lat),1]);
r_out = []; lon_plot_2 = []; lat_plot_2 = [];
depthrange = min_depth:1:max_depth;
for dd = depthrange
    lon_plot_2 = [lon_plot_2; gridded_lon(:)];
    lat_plot_2 = [lat_plot_2; gridded_lat(:)];
    temp = (s.Radius-dd*1e3)*ones(size(gridded_lon));
    r_out = [r_out; temp(:)];
end

visco_choice = 1;
if visco_choice == 0
    stress_deflection_input = 0;
    ref_system_input = 1;
    if stress_deflection_input == 0
        quantity = 'stresses';
    else
        quantity = 'deflections';
    end
    if ref_system_input == 0
        ref_system = 'cartesian';
    else
        ref_system = 'geographical';
    end
    diff_matrix_path = [diff_matrix_path_incomplete '\' quantity '\' ref_system];
    diff_plots_folder = [diff_matrix_path_incomplete '\plots\' quantity '\' ref_system];
else
    visco_strain_input = 0;
    if visco_strain_input == 0
        quantity = 'viscosity';
    else
        quantity = 'strain_rate';
    end
    diff_matrix_path = [diff_matrix_path_incomplete '\' quantity];
    diff_plots_folder = [diff_matrix_path_incomplete '\plots\' quantity];
end
dir_diff_files = dir([diff_matrix_path '\*.csv']);

if ~exist(diff_plots_folder, 'dir')
    mkdir(diff_plots_folder)
end
diff_files = {dir_diff_files.name};

matrix_1_flag = 0;
while matrix_1_flag == 0
%     quintuplet_1 = input(['Enter the iteration, step and cycle number, and the minimum and maximum depth '...
%     'of the first matrix used to plot the difference with respect to another moment in time '...
%     'as a vector of square brackets of five values:\n']);
%     for i=1:length(diff_files)
%         found_values_1 = str2double(regexp(diff_files{i}, '\d+', 'match'));
    for i=1:length(diff_files)
        quintuplet_1 = [str2double(iteration) str2double(step) 2 ...
            depths_to_plot(1) depths_to_plot(2)];
        found_values_1 = str2double(regexp(diff_files{i}, '\d+', 'match'));
        if isequal(quintuplet_1,found_values_1)
            matrix_1_flag = 1;
            file_to_read_1 = diff_files{i};
            iteration_1 = quintuplet_1(1);
            step_1 = quintuplet_1(2);
            cycle_1 = quintuplet_1(3);
            break;
        end
    end
    if matrix_1_flag == 1
        break;
    else
        disp('No match for the selected values, the matrix does not exist. Exiting search.');
        break;
    end
end

matrix_2_flag = 0;
while matrix_2_flag == 0
%     quintuplet_2 = input(['Enter the run, step and cycle number, and the minimum and maximum depth '...
%     'of the second matrix  used to plot the difference with respect to another moment in time '...
%     'as a vector of square brackets of five values:\n']);
%     for i=1:length(diff_files)
%         found_values_2 = str2double(regexp(diff_files{i}, '\d+', 'match'));
    for i=1:length(diff_files)
        quintuplet_2 = [str2double(iteration) str2double(step) 3 ...
            depths_to_plot(1) depths_to_plot(2)];
        found_values_2 = str2double(regexp(diff_files{i}, '\d+', 'match'));
        if isequal(quintuplet_2,found_values_2)
            matrix_2_flag = 1;
            file_to_read_2 = diff_files{i};
            iteration_2 = quintuplet_2(1);
            step_2 = quintuplet_2(2);
            cycle_2 = quintuplet_2(3);
            break;
        end
    end
    if matrix_2_flag == 1
        break;
    else
        disp('No match for the selected values, the matrix does not exist. Exiting search.');
        break;
    end
end
min_depth = quintuplet_2(4);
max_depth = quintuplet_2(5);

if matrix_1_flag == 1 && matrix_2_flag == 1
    % Extract the headers of both matrices, intersect and display the
    % possible common variables
    [values_1,headers_1,~] = xlsread([diff_matrix_path '\' file_to_read_1]);
    [values_2,headers_2,~] = xlsread([diff_matrix_path '\' file_to_read_2]);
    possible_variables_for_plotting = intersect(headers_1,headers_2, 'stable');
    possible_variables_for_plotting{end-2} = {}; 
    possible_variables_for_plotting{end-1} = {};
    possible_variables_for_plotting{end} = {};
    possible_variables_for_plotting = possible_variables_for_plotting(~cellfun('isempty',possible_variables_for_plotting));
    disp(['The possible variables to generate the difference plots between these'...
        ' two cycles are:\n']);
    disp(possible_variables_for_plotting);
    diff_variables_to_plot = possible_variables_for_plotting;
    filtered_lat = values_1(:,end-2);
    filtered_lon = values_1(:,end-1);
    depth_out = values_1(:,end);
    filtered_R = s.Radius - 1e3*depth_out;
    for i=1:length(diff_variables_to_plot)
        diff_variable_1 = values_1(:,strcmp(headers_1,diff_variables_to_plot{i}));
        diff_variable_2 = values_2(:,strcmp(headers_2,diff_variables_to_plot{i}));
        diff = diff_variable_2-diff_variable_1;
        [x_in,y_in,z_in]=sph2cart(deg2rad(filtered_lon),deg2rad(filtered_lat),filtered_R);
        [x_out,y_out,z_out]=sph2cart(deg2rad(lon_plot_2),deg2rad(lat_plot_2),r_out);
        plot_variable_out = griddata(x_in,y_in,z_in,diff,x_out,y_out,z_out,'nearest');
        
        figure(figure_counter)
        colormap summer;
        load coastlines;
        [Z, refvec] = geoloc2grid(lat_plot_2,lon_plot_2,plot_variable_out,resolution);
        ax = axesm('stereo','MapLatLimit',latlim,'MapLonLimit',lonlim,'Grid','on','MeridianLabel','on','ParallelLabel','on');
        set(ax,'Visible','off');
        cmap = colormap('summer');
        alpha 0.7;
        colormap(cmap);
        set(findall(gca, 'type', 'text'), 'visible', 'on')
        geoshow(Z, refvec, 'DisplayType', 'texture');
        plotm(coastlat, coastlon, 'color', rgb('OrangeRed'));
        caxis('auto'); % was [0 1e6]
        h = colorbar('v'); %set(h, 'ylim', [0 1e6]);
        if visco_choice == 1
            if visco_strain_input == 0
                set(get(h,'ylabel'),'string',[ diff_variables_to_plot{i} ' [N \cdot s/m^2]'])
            else
                set(get(h,'ylabel'),'string',[ diff_variables_to_plot{i}])
            end
        else
            if stress_deflection_input == 0
                set(get(h,'ylabel'),'string',[diff_variables_to_plot{i} ' [Pa]'])
            else
                set(get(h,'ylabel'),'string',[diff_variables_to_plot{i} ' [m]'])
            end
        end
%         latlim = [min_lat max_lat];lonlim = [min_lon max_lon];
%         ax = axesm('stereo','MapLatLimit',latlim,'MapLonLimit',lonlim,'Grid','on','MeridianLabel','on','ParallelLabel','on');
%         set(ax,'Visible','off');
%         set(findall(gca, 'type', 'text'), 'visible', 'on')
%         geoshow(Z, refvec, 'DisplayType', 'texture');
%         plotm(coastlat,coastlon);
        if visco_choice == 0
            title({['Map of the ' ref_system ' ' quantity ' difference for part with depth range '],...
                [num2str(min_depth) '-' num2str(max_depth) 'km and component ' diff_variables_to_plot{i} ','],...
                ['iterations [' num2str(iteration_1) '-' num2str(iteration_2) '], steps ['...
                num2str(step_1) '-' num2str(step_2) '], cycles [' num2str(cycle_1)...
                '-' num2str(cycle_2) ']']});
        else
            title({['Map of the ' quantity ' difference for part with depth range '],...
                [num2str(min_depth) '-' num2str(max_depth) 'km,' ],...
                ['iterations [' num2str(iteration_1) '-' num2str(iteration_2) '], steps ['...
                num2str(step_1) '-' num2str(step_2) '], cycles [' num2str(cycle_1)...
                '-' num2str(cycle_2) ']']}); 
        end
        set(findall(gca, 'type', 'text'), 'visible', 'on')
        grid on;
        % Save the figures based on whether we are working with
        % stresses or deflections
        if visco_choice == 1
            saveas(gcf,[diff_plots_folder '\diff_plot_' quantity '_'...
                '_depth_range_[' num2str(min_depth) '-' num2str(max_depth) ']_km_'...
                '_' run '_[' num2str(iteration_1)...
                '_' num2str(iteration_2) ']_[' num2str(step_1) '_' num2str(step_2)...
                ']_[' num2str(cycle_1) '_' num2str(cycle_2) '].png']);
        else
            saveas(gcf,[diff_plots_folder '\diff_plot_' ref_system '_' quantity '_'...
                '[' num2str(min_depth) '-' num2str(max_depth) ']_km_'...
                diff_variables_to_plot{i} '_' run '_[' num2str(iteration_1)...
                '_' num2str(iteration_2) ']_[' num2str(step_1) '_' num2str(step_2)...
                ']_[' num2str(cycle_1) '_' num2str(cycle_2) '].png']);
        end
        figure_counter = figure_counter + 1;
    end
end

end

File: python_files_python_3/diff_calculator.py
import os
import pdb
import numpy as np
import cartopy.crs as chart
import matplotlib.pyplot as plt
import scipy.interpolate as interp
import pandas as pd
import re
from matplotlib import cm
from plot_scale_extremes import *
from geo2cart import *
import math


def diff_calculator(diff_matrix_path_incomplete, min_lat, max_lat, min_lon, max_lon, run, depths_to_plot, iteration,
                    step, figure_counter):
    # [1 0 1 145 150]
    # [1 0 2 145 150]
    resolution = 0.25
    latlim = [min_lat, max_lat]
    lonlim = [min_lon, max_lon]
    min_depth = depths_to_plot[0]
    max_depth = depths_to_plot[1]
    earth_radius = 6371000
    lat_lin = np.arange(max_lat, min_lat, -resolution)
    lon_lin = np.arange(min_lon, max_lon, resolution)
    [gridded_lon, gridded_lat] = np.meshgrid(lon_lin, lat_lin)
    r_out = np.array([])
    lon_plot = np.array([])
    lat_plot = np.array([])
    depthrange = np.arange(min_depth, max_depth, 1)
    for dd in depthrange:
        lon_plot = np.vstack((lon_plot, gridded_lon)) if lon_plot.size else gridded_lon
        lat_plot = np.vstack((lat_plot, gridded_lat)) if lat_plot.size else gridded_lat
        temp = (earth_radius - dd * 1e3) * np.ones((len(gridded_lon), len(gridded_lon[0])))
        r_out = np.vstack([r_out, temp]) if r_out.size else temp

    visco_choice = 0
    if visco_choice == 0:
        stress_deflection_input = 1
        ref_system_input = 1
        if stress_deflection_input == 0:
            quantity = 'stresses'
        else:
            quantity = 'deflections'
        if ref_system_input == 0:
            ref_system = 'cartesian'
        else:
            ref_system = 'geographical'
        diff_matrix_path = os.path.join(diff_matrix_path_incomplete, quantity, ref_system)
        diff_plots_folder = os.path.join(diff_matrix_path_incomplete, 'plots', quantity, ref_system)
    else:
        visco_strain_input = 0
        if visco_strain_input == 0:
            quantity = 'viscosity'
        else:
            quantity = 'strain_rate'
        diff_matrix_path = os.path.join(diff_matrix_path_incomplete, quantity)
        diff_plots_folder = os.path.join(diff_matrix_path_incomplete, 'plots', quantity)

    dir_diff_files = os.path.join(diff_matrix_path, '*.csv')
    diff_files = []
    if not os.path.exists(diff_plots_folder):
        os.mkdir(diff_plots_folder)
    for i in range(len(dir_diff_files)):
        file_name = dir_diff_files[i].split('\\')
        diff_files.append(file_name[-1])
    matrix_1_flag = 0
    found_values_1 = []
    quintuplet_1 = np.arr([])
    while matrix_1_flag == 0:
        quintuplet_1 = input('Enter the iteration, step and cycle number, and the minimum and maximum depth of the first '
                                'matrix used to plot the difference with respect to another moment in time as a vector of '
                                'square brackets of five values: \n')
        for i in range(len(diff_files)):
            found_values_1 = re.findall(r"\((\d+(?:,\d+)*)\)", quintuplet_1)
            if (quintuplet_1 == found_values_1).all():
                matrix_1_flag = 1
                file_to_read_1 = diff_files[i]
                iteration_1 = quintuplet_1[0]
                step_1 = quintuplet_1[1]
                cycle_1 = quintuplet_1[2]
                break
        if matrix_1_flag == 1:
            break
        else:
            print('No match for the selected values, the matrix does not exist. Exiting search.')
            break

    found_values_2 = []
    matrix_2_flag = 0
    quintuplet_2 = np.arr([])
    while matrix_2_flag == 0:
        quintuplet_2 = input('Enter the iteration, step and cycle number, and the minimum and maximum depth of the first '
                                'matrix used to plot the difference with respect to another moment in time as a vector of '
                                'square brackets of five values: \n')
        for i in range(len(diff_files)):
            found_values_2 = re.findall(r"\((\d+(?:,\d+)*)\)", quintuplet_2)
            if (quintuplet_2 == found_values_2).all():
                matrix_1_flag = 1
                file_to_read_2 = diff_files[i]
                iteration_2 = quintuplet_2[0]
                step_2 = quintuplet_2[1]
                cycle_2 = quintuplet_2[2]
                break
        if matrix_2_flag == 1:
            break
        else:
            print('No match for the selected values, the matrix does not exist. Exiting search.')
            break

    if matrix_1_flag == 1 and matrix_2_flag == 1:
        min_depth = quintuplet_2[3]
        max_depth = quintuplet_2[4]
        headers_1 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_1), nrows = 1)
        headers_2 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_2), nrows = 1)
        values_1 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_1), skiprows = 1)
        values_2 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_2), skiprows = 1)
        possible_variables_for_plotting = headers_1.intersection(headers_2, 'stable')
        possible_variables_for_plotting[-3] = []
        possible_variables_for_plotting[-2] = []
        possible_variables_for_plotting[-1] = []
        print('The possible variables to generate the difference plots between these two cycles are: \n')
        print(possible_variables_for_plotting)
        diff_variables_to_plot = possible_variables_for_plotting
        filtered_lat = values_1[:, -3]
        filtered_lon = values_1[:, -2]
        depth_out = values_1[:, -1]
        filtered_r = earth_radius - 1e3 * depth_out
        for i in range(len(diff_variables_to_plot)):
            diff_variable_1 = values_1[:, (headers_1 == diff_variables_to_plot[i]).all()]
            diff_variable_2 = values_2[:, (headers_2 == diff_variables_to_plot[i]).all()]
            plot_variable = diff_variable_2 - diff_variable_1
            [x_in, y_in, z_in] = geo2cart(filtered_lat, filtered_lon, filtered_r)
            [x_out, y_out, z_out] = geo2cart(np.deg2rad(lon_plot), np.deg2rad(lat_plot), r_out)
            plot_variable_out = interp.griddata(x_in, y_in, z_in, plot_variable, x_out, y_out, z_out, 'nearest')

            # Open image
            diff_figure = plt.figure()
            # Geographic map
            ax = plt.axes(projection=chart.Orthographic(0, -90))
            ax.coastlines(resolution='50m', color='orange', linewidth=1)
            # Add surface
            ax.plot_surface(lon_plot, lat_plot, plot_variable_out, cmap=cm.summer, linewidth=0, antialiased=False)
            # Colorbar settings
            scale = plt.colorbar(diff_figure)
            diff_figure.clim('auto')
            scale = plt.colorbar(diff_figure)
            if visco_choice == 1:
                if visco_strain_input == 0:
                    scale.set_label(diff_variables_to_plot[i] + ' [N \cdot s/m^2]')
                else:
                    scale.set_label(diff_variables_to_plot[i])
            else:
                if stress_deflection_input == 0:
                    scale.set_label(diff_variables_to_plot[i] + ' [Pa]')
                else:
                    scale.set_label(diff_variables_to_plot[i] + ' [m]')
            if visco_choice == 0:
                if diff_variables_to_plot[i] == 'Mises':
                    plt.title('Map of value differences for the Mises stresses')
                elif diff_variables_to_plot[i] == 'Magnitude':
                    plt.title('Map of value differences for the deformation magnitude')
                else:
                    plt.title('Map of value differences for ' +  quantity + ' component ' + diff_variables_to_plot[i])
            else:
                plt.title('Map of differences for the ' + quantity)
            if visco_choice == 1:
                plt.savefig(os.path.join(diff_plots_folder, 'diff_plot_' + quantity + '_' + '_depth_range_[' + str(min_depth)
                + '-' + str(max_depth) + ']_km_' + '_' + run + '_[' + str(iteration_1) + '_' + str(iteration_2) + ']_['
                + str(step_1) + '_' + str(step_2) + ']_[' + str(cycle_1) + '_' + str(cycle_2) + '].png'))
            else:
                plt.savefig(os.path.join(diff_plots_folder, 'diff_plot_' + quantity + '_' + '_depth_range_[' + str(min_depth)
                + '-' + str(max_depth) + ']_km_' + '_' + run + '_[' + str(iteration_1) + '_' + str(iteration_2) + ']_['
                + str(step_1) + '_' + str(step_2) + ']_[' + str(cycle_1) + '_' + str(cycle_2) + '].png'))

        figure_counter = figure_counter + 1